src/utils/allocator_default.h

```diff
@@ -35,4 +39,16 @@ static inline void *omf_realloc_real(void *ptr, size_t size, const char *file, i
     abort();
 }
 
+static inline void *omf_alloc_with_options_real(size_t nmemb, size_t size, int options, const char *file, int line) {
```

Comment:
Do we need the options ? Could this just be omF_alloc_surface_data() or somesuch ?

Reply:
It could, though keeping it more abstract allows us to reduce the number of functions in the API and changes don't need to touch the common code if adding new options. It also allows implementations to ignore new options when they are added, instead of having to implement the missing functions.

Reply:
I have three issues with that:
1. We don't currently know of any cases that need more flags, making this an unnecessary complication
2. Free function also requires the same flags as the alloc, which us ugly
3. Why do we need a bitmask ? malloc cases tend to be pretty specific anyways, making separate functions clearer in the long run.

Reply:
> 1. We don't currently know of any cases that need more flags, making this an unnecessary complication

Any allocation that could be used by hardware directly, will need to go through alloc with options.

> 2. Free function also requires the same flags as the alloc, which us ugly

This is more of a stylistic preference why is a set of flags and 1 free-function uglier than a set of free-functions names.

> 3. Why do we need a bitmask ? malloc cases tend to be pretty specific anyways, making separate functions clearer in the long run.

Allocation parameters may be different, cached/uncached, aligned/unaligned, size-extended/non-extended, different alginment requirements for different allocation types.
My gripe with separate functions is that you will have to rev all your platforms every time a new function is added and most platforms won't do anything special or different. So just increasing the merge conflict size, when multiple people are working on that code.

Reply:
bitflags are unsifficient for arbitrarily aligned allocations and allocate_at_least style allocations.

Arbitrarily (un)aligned allocations do not map particularly well to bitflags, as the caller needs to control to what the allocation is aligned-- 32-byte alignment for AVX, 16-byte for SSE, etc.. I'd expect any aligned allocator we provide to be similar to [aligned_alloc](https://en.cppreference.com/w/c/memory/aligned_alloc).

size-extended allocations.. sounds like when vector.h asks for an allocation of size 12 (for 3 4-byte elements), and the allocator decides to round up to some common size like 16-- the allocator must have a way of passing the size of the allocation to an interested caller (see C++'s [allocate_at_least](https://en.cppreference.com/w/cpp/memory/allocator_traits/allocate_at_least)), preferrably without the overhead of storing alloc size in every allocation (as glibc does for malloc_usable_size).

Reply:
> bitflags are unsifficient for arbitrarily aligned allocations and allocate_at_least style allocations.

I've thought about this too and initially thought it wouldn't be sufficient but warmed up to the idea of letting the platform decide on the allocation type based on the incoming flags. 
The flags are very high level but allow the platform to make a decision if it should be any of the above types. While alloc_aligned may make sense on one platform it doesn't have to make sense on another and can be unaligned somewhere else.